File: leetcode-go/Q001-Q100/Q044_Wildcard_Matching/Q044_Wildcard_Matching.go
package Q044_Wildcard_Matching

func isMatch(s string, p string) bool {
	dp := make([][]bool, len(p)+1)
	for i := 0; i < len(dp); i++ {
		dp[i] = make([]bool, len(s)+1)
	}
	dp[len(p)][len(s)] = true

	for i := len(p) - 1; i >= 0; i-- {
		if p[i] != '*' {
			break
		}
		dp[i][len(s)] = true
	}

	for i := len(p) - 1; i >= 0; i-- {
		for j := len(s) - 1; j >= 0; j-- {
			if p[i] == s[j] || p[i] == '?' {
				dp[i][j] = dp[i+1][j+1]
			} else if p[i] == '*' {
				dp[i][j] = dp[i+1][j] || dp[i][j+1]
			} else {
				dp[i][j] = false
			}
		}
	}
	return dp[0][0]
}

File: src/Y2018/facebook/dp/Q044_Wildcard_Matching.java
package Y2018.facebook.dp;

public class Q044_Wildcard_Matching {

    public boolean isMatch(String s, String p) {
        char[] ss = s.toCharArray(), pp = p.toCharArray();
        boolean[][] dp = new boolean[ss.length + 1][pp.length + 1];
        dp[0][0] = true;
        for (int j = 0; j < pp.length; j++) {
            if (pp[j] == '*') {
                dp[0][j + 1] = dp[0][j];
            }
        }
        for (int i = 0; i < ss.length; i++) {
            for (int j = 0; j < pp.length; j++) {
                if (ss[i] == pp[j] || pp[j] == '?') {
                    dp[i + 1][j + 1] |= dp[i][j];
                }
                if (pp[j] == '*') {
                    dp[i + 1][j + 1] |= dp[i - 1][j] || dp[i][j - 1];
                }
            }
        }
        return dp[ss.length][pp.length];
    }

}
